src/level.c

```diff
@@ -0,0 +1,158 @@
+#include <assert.h>
+#include <SDL2/SDL.h>
+
+#include "./lt.h"
+#include "./player.h"
+#include "./platforms.h"
+#include "./level.h"
+#include "./camera.h"
+#include "./error.h"
+
+struct level_t
+{
+    lt_t *lt;
+    player_t *player;
+    platforms_t *platforms;
+};
+
+level_t *create_level(player_t *player,
+                      platforms_t *platforms)
+{
+    assert(player);
+    assert(platforms);
+
+    lt_t *const lt = create_lt();
+    if (lt == NULL) {
+        return NULL;
+    }
+
+    level_t *const level = PUSH_LT(lt, malloc(sizeof(level_t)), free);
+    if (level == NULL) {
+        throw_error(ERROR_TYPE_LIBC);
+        RETURN_LT(lt, NULL);
+    }
+
+    level->player = PUSH_LT(lt, player, destroy_player);
+    level->platforms = PUSH_LT(lt, platforms, destroy_platforms);
+    level->lt = lt;
+
+    return level;
+}
+
+level_t *create_level_from_file(const char *file_name)
+{
+    assert(file_name);
+
+    lt_t *const lt = create_lt();
+    if (lt == NULL) {
+        return NULL;
+    }
+
+    level_t *const level = PUSH_LT(lt, malloc(sizeof(level_t)), free);
+    if (level == NULL) {
+        throw_error(ERROR_TYPE_LIBC);
+        RETURN_LT(lt, NULL);
+    }
+
+    FILE *level_file = PUSH_LT(lt, fopen(file_name, "r"), fclose);
+    if (level_file == NULL) {
+        throw_error(ERROR_TYPE_LIBC);
+        RETURN_LT(lt, NULL);
+    }
+
+    level->player = PUSH_LT(lt, create_player_from_stream(level_file), destroy_player);
+    if (level->player == NULL) {
+        RETURN_LT(lt, NULL);
+    }
+
+    level->platforms = PUSH_LT(lt, create_platforms_from_stream(level_file), destroy_platforms);
+    if (level->platforms == NULL) {
+        RETURN_LT(lt, NULL);
+    }
+
+    level->lt = lt;
+
+    fclose(RELEASE_LT(lt, level_file));
+
+    return level;
+}
+
+void destroy_level(level_t *level)
+{
+    assert(level);
+    RETURN_LT0(level->lt);
+}
+
+int level_render(const level_t *level, camera_t *camera, SDL_Renderer *renderer)
+{
+    assert(level);
+    assert(renderer);
+
+    if (render_player(level->player, renderer, camera) < 0) {
+        return -1;
+    }
+
+    if (render_platforms(level->platforms, renderer, camera) < 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int level_update(level_t *level, camera_t *camera, Uint32 delta_time)
```

Comment:
Error handling in this functions is kinda strange. We say we return int, but never actually use it for error reporting. Revise this function.